### PR TITLE
gh-110222: Add support of PyStructSequence in copy.replace()

### LINKED ISSUE #110222: Add support of PyStructSequence in copy.replace()
# Feature or enhancement

### Proposal:

This issue is to make the concept of ["named tuple"](https://docs.python.org/3/glossary.html#term-named-tuple) support the `__replace__` protocol.

`collections.namedtuple` and `typing.NamedTuple` already support the `__replace__` protocol in:

- #108752

[`PyStructSequence`](https://docs.python.org/3/c-api/tuple.html#struct-sequence-objects)s are also [named tuples](https://docs.python.org/3/glossary.html#term-named-tuple) but they do not support the `__replace__` protocol yet.

It would be convenient if `PyStructSequence` be supported in `copy.replace()`.

### Has this already been discussed elsewhere?

I have already discussed this feature proposal on Discourse

### Links to previous discussion of this feature:

- #108751
- https://discuss.python.org/t/generalize-replace-function/28511
- #109956
- #109174
- #109175

### Linked PRs
* gh-110223

### PR DESCRIPTION
**Pull request by XuehaiPan**
Resolves #110222

~~The implementation in this PR is roughly equivalent to:~~

```python
class SomeStructSeq:
    def __replace__(self, /, **kwargs):
        if self.n_unamed_fields > 0:
            raise TypeError(f'__replace__() is not supported for {self.__class__} '
                            'because it has unnamed field(s)')

        field_names = self._get_field_names()
        _, (tup, dct) = self.__reduce__()
        seq = list(tup)
        for i, name in field_names[:self.n_sequence_fields]:
            if name in kwargs:
                seq[i] = kwargs.pop(name)
        for name in field_names[self.n_sequence_fields:]:
            if name in kwargs:
                dct[name] = kwargs.pop(name)
        if kwargs:
            raise TypeError(f'Got unexpected field name(s): {list(kwargs)!r}')
        return self.__class__(seq, dct)
```

Refactored:

```python
class SomeStructSeq:
    def __replace__(self, /, **kwargs):
        if self.n_unamed_fields > 0:
            raise TypeError(f'__replace__() is not supported for {self.__class__} '
                            'because it has unnamed field(s)')

        clone = self._clone()
        field_names = self._get_field_names()
        if kwargs:
            for i, name in enumerate(field_names):
                if name in kwargs:
                    clone[i] = kwargs.pop(name)
            if kwargs:
                raise TypeError(f'Got unexpected field name(s): {list(kwargs)!r}')
        return clone
```

* Issue: gh-110222

### CODE REVIEW

File: Lib/test/test_copy.py [hunk 1]
```diff
@@ -945,6 +947,28 @@ def test_namedtuple(self):
         with self.assertRaisesRegex(ValueError, 'unexpected field name'):
             copy.replace(p, x=1, error=2)
 
+    def test_structseq(self):
```
Comment (serhiy-storchaka) [inline Lib/test/test_copy.py]: I think that it is better to test it in `test_structseq` module. Most `__replace__` implementations are tested together with other methods of their class. Only exceptions are `namedtuple` and `dataclass` because they are fabrics to create new classes.

It may be worth to split the test on two tests: with and without unnamed fields.

File: Objects/structseq.c [hunk 2]
```diff
@@ -365,9 +366,99 @@ structseq_reduce(PyStructSequence* self, PyObject *Py_UNUSED(ignored))
...
+            }
+            PyObject *ob = _PyDict_Pop(kwargs, key, self->ob_item[i]);
+            if (PyDict_SetItem(dict, key, ob) < 0) {
+                goto error;
+            }
+        }
+        if (PyDict_Size(kwargs) > 0) {
+            PyObject *names = PyDict_Keys(kwargs);
+            if (names) {
+                PyErr_Format(PyExc_ValueError, "Got unexpected field name(s): %R", names);
```
Comment (serhiy-storchaka) [inline Objects/structseq.c]: I think that TypeError is more appropriate exception.

Yes, `namedtuple` raises ValueError, but it is unusual. Perhaps it should be changed.

Comment (XuehaiPan) [reply Objects/structseq.c]: > `namedtuple` raises ValueError

`dataclass.__replace__` also raises `ValueError`.

https://github.com/python/cpython/blob/6387b5313c60c1403785b2245db33372476ac304/Lib/dataclasses.py#L1555-L1586

Comment (serhiy-storchaka) [reply Objects/structseq.c]: It is only for InitVar fields. TypeError is raised for unknown names.

Comment (XuehaiPan) [reply Objects/structseq.c]: Thanks for the clarification. I update the exception type to `TypeError` for both `PyStructSequence` and `collections.namedtuple`.

File: Lib/collections/__init__.py [hunk 3]
```diff
@@ -457,7 +457,7 @@ def _make(cls, iterable):
     def _replace(self, /, **kwds):
         result = self._make(_map(kwds.pop, field_names, self))
         if kwds:
-            raise ValueError(f'Got unexpected field names: {list(kwds)!r}')
+            raise TypeError(f'Got unexpected field name(s): {list(kwds)!r}')
```
Comment (serhiy-storchaka) [inline Lib/collections/__init__.py]: Please revert this change and open a new issue for it. It is a breaking change and should be discussed and separately documented.

**Review by serhiy-storchaka — COMMENTED** [review]
I am surprised to see a correct complex C code from a new contributor. It is not easy to write all these error checks and decrefs on the first version.

The implementation creates arguments (a tuple and a dict) for `__new__()` and calls it. It is not the most efficient way, because it creates intermediate tuple and dict and can create string keys for name-only fields multiple times, but it works, and performance is not critical. I'm just wondering, whether it was your first idea or you considered other approaches and they turned out more complex? Did you try to create a new instance with `PyStructSequence_New()`, then copy `ob_items` from old structure to the new, then iterate the kwargs dict in one pass and patch `ob_items`? It looks simpler to me, but I may have missed some details.

**Comment by XuehaiPan** [comment]
> I'm just wondering, whether it was your first idea or you considered other approaches and they turned out more complex? Did you try to create a new instance with `PyStructSequence_New()`, then copy `ob_items` from old structure to the new, then iterate the kwargs dict in one pass and patch `ob_items`? It looks simpler to me, but I may have missed some details.

@serhiy-storchaka My first thought is to create a copy of the current structseq and then replace the new values from `kwargs`. But I found it is kind of complex to handle the unnamed fields. Then I implement it with a similar structure to the `__reduce__()` implementation.

I will refactor it with the former implementation since we have disabled the support for types with unnamed fields.

**Comment by XuehaiPan** [comment]
> I will refactor it with the former implementation since we have disabled the support for types with unnamed fields.

Done.

File: Lib/collections/__init__.py [hunk 4]
```diff
@@ -457,7 +457,7 @@ def _make(cls, iterable):
     def _replace(self, /, **kwds):
         result = self._make(_map(kwds.pop, field_names, self))
         if kwds:
-            raise ValueError(f'Got unexpected field names: {list(kwds)!r}')
+            raise ValueError(f'Got unexpected field name(s): {list(kwds)!r}')
```
Comment (serhiy-storchaka) [inline Lib/collections/__init__.py]: It is not related change.

File: Lib/test/test_structseq.py [hunk 5]
```diff
@@ -133,6 +135,74 @@ def test_match_args_with_unnamed_fields(self):
         self.assertEqual(os.stat_result.n_unnamed_fields, 3)
         self.assertEqual(os.stat_result.__match_args__, expected_args)
 
+    def test_copy_replace_all_fields_visible(self):
+        assert os.times_result.n_unnamed_fields == 0 and os.times_result.n_sequence_fields == os.times_result.n_fields
+
+        expected_args = ('user', 'system', 'children_user', 'children_system', 'elapsed')
+        self.assertEqual(os.times_result.__match_args__, expected_args)
+        self.assertEqual(os.times_result.n_fields, len(expected_args))
```
Comment (serhiy-storchaka) [inline Lib/test/test_structseq.py]: I am not sure it is related here.

File: Lib/test/test_structseq.py [hunk 6]
```diff
@@ -133,6 +135,74 @@ def test_match_args_with_unnamed_fields(self):
...
+
+        n_fields = os.times_result.n_fields  # 5
+        t = os.times_result(range(os.times_result.n_fields))
+
+        # visible fields
+        self.assertEqual(copy.replace(t), tuple(range(n_fields)))
+        self.assertEqual(copy.replace(t, user=1), (1, *range(1, n_fields)))
+        self.assertEqual(copy.replace(t, system=2), (0, 2, *range(2, n_fields)))
+        self.assertEqual(copy.replace(t, user=1, system=2), (1, 2, *range(2, n_fields)))
```
Comment (serhiy-storchaka) [inline Lib/test/test_structseq.py]: Few suggestions:
* Test that the result has the correct type, not simple tuple.
* Use float values. Currently there are no restrictions, but normally they are floats.
* Use subscription and slices to get original field values.

```suggestion
        self.assertEqual(copy.replace(t), t)
        self.assertIsInstance(copy.replace(t), os.times_result)
        self.assertEqual(copy.replace(t, user=1.5), (1.5, *t[1:]))
        self.assertEqual(copy.replace(t, system=2.5), (t[0], 2.5, *t[2:]))
        self.assertEqual(copy.replace(t, user=1.5, system=2.5), (1.5, 2.5, *t[2:]))
```

File: Lib/test/test_structseq.py [hunk 7]
```diff
@@ -133,6 +135,74 @@ def test_match_args_with_unnamed_fields(self):
...
 
+    def test_copy_replace_all_fields_visible(self):
+        assert os.times_result.n_unnamed_fields == 0 and os.times_result.n_sequence_fields == os.times_result.n_fields
+
+        expected_args = ('user', 'system', 'children_user', 'children_system', 'elapsed')
+        self.assertEqual(os.times_result.__match_args__, expected_args)
+        self.assertEqual(os.times_result.n_fields, len(expected_args))
+
+        n_fields = os.times_result.n_fields  # 5
+        t = os.times_result(range(os.times_result.n_fields))
```
Comment (serhiy-storchaka) [inline Lib/test/test_structseq.py]: You can simply use `os.times()`. It guarantees that the result is correct.

File: Lib/test/test_structseq.py [hunk 8]
```diff
@@ -133,6 +135,74 @@ def test_match_args_with_unnamed_fields(self):
...
+            copy.replace(t, error=-1)
+        with self.assertRaisesRegex(TypeError, 'unexpected field name'):
+            copy.replace(t, user=1, error=-1)
+
+    def test_copy_replace_with_invisible_fields(self):
+        t = time.gmtime(0)
+        assert t.n_unnamed_fields == 0 and t.n_sequence_fields < t.n_fields
+
+        # visible fields
+        self.assertEqual(copy.replace(t), (1970, 1, 1, 0, 0, 0, 3, 1, 0))
```
Comment (serhiy-storchaka) [inline Lib/test/test_structseq.py]: Please test also the type and "invisible" attributes of the result.
```suggestion
        t2 = copy.replace(t)
        self.assertEqual(t2, (1970, 1, 1, 0, 0, 0, 3, 1, 0))
        self.assertIsInstance(t2, time.struct_time)
        self.assertEqual(t2.tm_zone, 'GMT')
```

File: Objects/structseq.c [hunk 9]
```diff
@@ -365,9 +366,82 @@ structseq_reduce(PyStructSequence* self, PyObject *Py_UNUSED(ignored))
...
+        PyErr_Format(PyExc_TypeError,
+                     "__replace__() is not supported for %.500s "
+                     "because it has unnamed field(s)",
+                     Py_TYPE(self)->tp_name);
+        return NULL;
+    }
+
+    result = (PyStructSequence *) PyStructSequence_New(Py_TYPE(self));
+    if (!result) {
+        goto error;
```
Comment (serhiy-storchaka) [inline Objects/structseq.c]: It can be simply `return NULL`.

File: Objects/structseq.c [hunk 10]
```diff
@@ -365,9 +366,82 @@ structseq_reduce(PyStructSequence* self, PyObject *Py_UNUSED(ignored))
...
+
+    if (kwargs != NULL) {
+        // We do not support types with unnamed fields, so we can iterate over
+        // i >= n_visible_fields case without slicing with (i - n_unnamed_fields).
+        for (i = 0; i < n_fields; ++i) {
+            PyObject *key = PyUnicode_FromString(Py_TYPE(self)->tp_members[i].name);
+            if (!key) {
+                goto error;
+            }
+            PyObject *ob = _PyDict_Pop(kwargs, key, self->ob_item[i]);  // borrowed
```
Comment (serhiy-storchaka) [inline Objects/structseq.c]: It is not borrowed.

**Review by serhiy-storchaka — COMMENTED** [review]
Very well.

I have few more comments, mostly test suggestions.

File: Lib/test/test_structseq.py [hunk 11]
```diff
@@ -133,6 +135,88 @@ def test_match_args_with_unnamed_fields(self):
         self.assertEqual(os.stat_result.n_unnamed_fields, 3)
         self.assertEqual(os.stat_result.__match_args__, expected_args)
 
+    def test_copy_replace_all_fields_visible(self):
+        assert (
+            os.times_result.n_unnamed_fields == 0
+            and os.times_result.n_sequence_fields == os.times_result.n_fields
+        )
```
Comment (serhiy-storchaka) [inline Lib/test/test_structseq.py]:
```suggestion
        assert os.times_result.n_unnamed_fields == 0
        assert os.times_result.n_sequence_fields == os.times_result.n_fields
```

File: Lib/test/test_structseq.py [hunk 12]
```diff
@@ -133,6 +135,88 @@ def test_match_args_with_unnamed_fields(self):
...
+        with self.assertRaisesRegex(TypeError, 'unexpected field name'):
+            copy.replace(t, error=-1)
+        with self.assertRaisesRegex(TypeError, 'unexpected field name'):
+            copy.replace(t, user=1, error=-1)
+
+    def test_copy_replace_with_invisible_fields(self):
+        assert (
+            time.struct_time.n_unnamed_fields == 0
+            and time.struct_time.n_sequence_fields < time.struct_time.n_fields
+        )
```
Comment (serhiy-storchaka) [inline Lib/test/test_structseq.py]:
```suggestion
        assert time.struct_time.n_unnamed_fields == 0
        assert time.struct_time.n_sequence_fields < time.struct_time.n_fields
```

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
LGTM!

**Comment by serhiy-storchaka** [comment]
Thank you for your contribution @XuehaiPan.
